Add logging for initial publish of a course

alexsanford opened the pull request:
Closes #2650 

Adds logging for when a course is initially published but does not log if the course is unpublished and published again later.

## Implementation note

There is a check here for meta that is added by WCPC. I don't really like it being in this codebase since it seems like it should be in WCPC itself. However, since it is just meta, there isn't any strong technical reason why this plugin couldn't simply check for it. This keeps it much simpler but introduces a funny dependency.

Open to thoughts on this.

## Testing instructions

_Note: please test this in both the classic editor **and** the block editor._

- Create a course and save it as a draft.
- Add modules to the course.
- Add lessons to the course.
- Publish the course. Ensure that the logged event has the correct attributes (including `-1` for the `product_id`).
- Unpublish the course and publish it again. Ensure that another event is not logged.
- Follow steps 1-3 and then install WCPC and link a product to the course.
- Publish the course, and ensure the `product_id` attribute on the event is correct.
- Create another draft course, but do not include any modules or lessons.
- Publish the course and ensure that the `module_count` and `lesson_count` attributes are `0`.
- Create a course and add modules, lessons, and a product without saving as a draft (set all of these things while on the initial "Add Course" screen). On publishing, ensure that the event is logged, and the properties are set correctly.

### Additional test for existing installations

- On a site with courses that have already been published, switch to this branch.
- Unpublish and publish an existing course.
- Ensure that no event is logged.

jom commented on includes/class-sensei-course.php:
```diff
@@ -3313,6 +3316,44 @@ public static function prerequisite_complete_message() {
+	 * @param WP_Post $post       The course.
+	 */
+	public function log_initial_publish_event( $new_status, $old_status, $post ) {
+		if ( 'course' !== $post->post_type ) {
```
You should add a check for `|| $new_status === $old_status` here. This hook is called even for `publish -> publish`

alexsanford commented:
Good catch @jom! That never even occurred to me. Should be fixed in c644d3f.

Also fixed linter issues in 56d5d89. (Note that I also rebased to get the linter issues fixed on `add/logging-settings-save`)

As for the case where a post was published and then moved to a draft before this update, I don't think there is any way to handle that, since we have no way (to my knowledge) of determining the status history of a post. So I think we only have two options: log the event for all unpublished courses that exist before this update, or do not log the event for any of them. Currently, this PR will log for all of them, but I'm open to doing it the other way instead.

donnapep commented:
@alexsanford Is this one ready for a review, or should I wait a bit?

alexsanford commented:
@donnapep It can be reviewed now. I just have it in draft until #2684 is merged since it uses the testing framework from that PR. I don't anticipate any large changes to #2684 that will affect this.

donnapep commented:
> As for the case where a post was published and then moved to a draft before this update, I don't think there is any way to handle that

When I tried this it didn't log it. Did it log it for you?

donnapep commented on includes/class-sensei-course.php:
```diff
@@ -3313,6 +3316,44 @@ public static function prerequisite_complete_message() {
+
+		// If we are publishing or unpublishing, mark as already published.
+		if ( ( $publishing || $unpublishing ) && ! $already_published ) {
+			add_post_meta( $post->ID, 'course_already_published', true, true );
```
Just curious if this post meta will be deleted when the plugin is uninstalled?

alexsanford replied:
Good question!

Since this is post meta being added to a course, it will be deleted on uninstall (if the user has selected that option). All post meta is deleted when its post is deleted, and all courses are deleted on uninstall.

The only caveat is that the courses may be moved to trash first, and then permanently deleted after the configured amount of time (default 30 days). In which case the meta will not be deleted until the course is permanently deleted.

alexsanford commented:
I found a pretty serious bug in this PR, and so I'm moving it to a WIP.

In short, everything works fine when saving a course as a draft and then publishing later. But if you publish right away without saving a draft first, the properties are not logged correctly.

I've added an additional testing instruction to the description above to catch this:

> Create a course and add modules, lessons, and a product without saving as a draft (set all of these things while on the initial "Add Course" screen). On publishing, ensure that the event is logged, and the properties are set correctly.

alexsanford commented:
@donnapep @jom Ok, I've implemented a fix for the issue described above (thanks @jom for the help!)

I've re-requested review because I've changed the implementation pretty substantially. If you could have another look, and re-test, that would be great. The testing instructions are still as per the description, with the additional point that I added previously.

## Some implementation notes

Along with fixing the described issue, I also added some changes to make this more generic. This will let us use this functionality easily for Lessons, and any other CPT's that we may want to track.

The code is a bit confusing, and there are a few edge cases I'm dealing with. I tried to add helpful documentation comments, but please let me know if anything is still unclear.

donnapep commented:
@alexsanford For the case where the course is published without going to Draft first, if there is no product attached, it's set to `-` instead of `-1`.

donnapep commented:
@alexsanford Also, if I add a course via the Lessons page, the Product ID isn't being picked up. It's always set to `-1`.

alexsanford commented:
> For the case where the course is published without going to Draft first, if there is no product attached, it's set to - instead of -1.

Fixed in 086c619

> Also, if I add a course via the Lessons page, the Product ID isn't being picked up. It's always set to -1.

Oof. Another big problem that required a refactor. Not as bad as last time, though.

This should be fixed in 78897b7. Please re-test 🙂 

@jom Do you mind having a look at the refactor, to see if it looks sensible? It's basically what you and I chatted about: adding the post ID's to an array, and looping through them in the `shutdown` hook to fire the actions.

alexsanford commented:
The build is fixed, although it uncovered another issue with CI. I've outlined the new issue in https://github.com/Automattic/sensei/issues/2710. But I think we can go ahead with merging this one now.
